# Project : Analysing Music Records Sales using SQL


Chinook is a fictional music record store based out of the USA. The store offers a wide range of music genres, from classic rock and jazz to modern indie and hip-hop. With a passion for music and a commitment to providing exceptional customer
experiences, Chinook aims to be the go-to destination for music lovers.


# To achieve so, they have the following business objectives:

1. Establish Chinook as the premier destination for music records.
2. Provide a diverse selection of music genres to cater to a wide range of customer preferences.
3. Deliver excellent customer service to enhance the shopping experience.
4. Build a loyal customer base and foster long-term relationships through engagement and promotional events.


In [2]:
import numpy as np 
import pandas as pd 
import sqlite3
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

database = 'chinook.db'

In [3]:
conn = sqlite3.connect(database)
c= conn.cursor()

tables = pd.read_sql("""SELECT *
                        FROM sqlite_master
                        WHERE type='table';""", conn)
tables

type            name        tbl_name  rootpage  \
0   table           album           album         2   
1   table          artist          artist         3   
2   table        customer        customer         4   
3   table        employee        employee         5   
4   table           genre           genre         6   
5   table         invoice         invoice         7   
6   table    invoice_line    invoice_line         8   
7   table      media_type      media_type         9   
8   table        playlist        playlist        10   
9   table  playlist_track  playlist_track        11   
10  table           track           track        13   

                                                  sql  
0   CREATE TABLE [album]\n(\n    [album_id] INTEGE...  
1   CREATE TABLE [artist]\n(\n    [artist_id] INTE...  
2   CREATE TABLE [customer]\n(\n    [customer_id] ...  
3   CREATE TABLE [employee]\n(\n    [employee_id] ...  
4   CREATE TABLE [genre]\n(\n    [genre_id] INTEGE...  
5   CREATE TABLE [invoice]\n(\n    [invoice_id] IN...  
6   CREATE TABLE [invoice_line]\n(\n    [invoice_l...  
7   CREATE TABLE [media_type]\n(\n    [media_type_...  
8   CREATE TABLE [playlist]\n(\n    [playlist_id] ...  
9   CREATE TABLE [playlist_track]\n(\n    [playlis...  
10  CREATE TABLE [track]\n(\n    [track_id] INTEGE...

# Segment 1: Database - Tables, Columns, Relationships

Identify the tables in the database and their respective columns

Determine the number of records in each table within the schema

In [4]:
query1 = pd.read_sql("""
SELECT *
FROM album;
""", conn)

query1

album_id                                              title  artist_id
0           1              For Those About To Rock We Salute You          1
1           2                                  Balls to the Wall          2
2           3                                  Restless and Wild          2
3           4                                  Let There Be Rock          1
4           5                                           Big Ones          3
..        ...                                                ...        ...
342       343                             Respighi:Pines of Rome        226
343       344  Schubert: The Late String Quartets & String Qu...        272
344       345                                Monteverdi: L'Orfeo        273
345       346                              Mozart: Chamber Music        274
346       347  Koyaanisqatsi (Soundtrack from the Motion Pict...        275

[347 rows x 3 columns]

- The columns in the DataFrame are 'album_id', 'title', and 'artist_id'. Each row represents a different album, and the columns provide specific details about each album.
- Each album is identified by its 'album_id' and has a corresponding 'title' representing the name of the album.
- The 'artist_id' column indicates the ID of the artist or band associated with each album.

In [6]:
query1 = pd.read_sql("""
SELECT *
FROM artist;
""", conn)

query1

artist_id                                               name
0            1                                              AC/DC
1            2                                             Accept
2            3                                          Aerosmith
3            4                                  Alanis Morissette
4            5                                    Alice In Chains
..         ...                                                ...
270        271   Mela Tenenbaum, Pro Musica Prague & Richard Kapp
271        272                             Emerson String Quartet
272        273  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...
273        274                                      Nash Ensemble
274        275                              Philip Glass Ensemble

[275 rows x 2 columns]

- The given data represents a DataFrame with information about various music artists.
- The DataFrame consists of two columns: 'artist_id' and 'name'. Each row corresponds to a different artist, and the columns provide specific details about each artist.
- The Music Artists DataFrame contains details about a diverse collection of music artists.
- Each artist is uniquely identified by their 'artist_id'. The 'name' column presents the name of the artist or band associated with their music

In [7]:
query1 = pd.read_sql("""
SELECT *
FROM customer;
""", conn)

query1

customer_id first_name     last_name  \
0             1       Luís     Gonçalves   
1             2     Leonie        Köhler   
2             3   François      Tremblay   
3             4      Bjørn        Hansen   
4             5  František   Wichterlová   
5             6     Helena          Holý   
6             7     Astrid        Gruber   
7             8       Daan       Peeters   
8             9       Kara       Nielsen   
9            10    Eduardo       Martins   
10           11  Alexandre         Rocha   
11           12    Roberto       Almeida   
12           13   Fernanda         Ramos   
13           14       Mark       Philips   
14           15   Jennifer      Peterson   
15           16      Frank        Harris   
16           17       Jack         Smith   
17           18   Michelle        Brooks   
18           19        Tim         Goyer   
19           20        Dan        Miller   
20           21      Kathy         Chase   
21           22    Heather       Leacock   
22           23       John        Gordon   
23           24      Frank       Ralston   
24           25     Victor       Stevens   
25           26    Richard    Cunningham   
26           27    Patrick          Gray   
27           28      Julia       Barnett   
28           29     Robert         Brown   
29           30     Edward       Francis   
30           31     Martha          Silk   
31           32      Aaron      Mitchell   
32           33      Ellie      Sullivan   
33           34       João     Fernandes   
34           35   Madalena       Sampaio   
35           36     Hannah     Schneider   
36           37       Fynn    Zimmermann   
37           38     Niklas      Schröder   
38           39    Camille       Bernard   
39           40  Dominique      Lefebvre   
40           41       Marc        Dubois   
41           42      Wyatt        Girard   
42           43   Isabelle       Mercier   
43           44      Terhi    Hämäläinen   
44           45   Ladislav        Kovács   
45           46       Hugh      O'Reilly   
46           47      Lucas       Mancini   
47           48   Johannes  Van der Berg   
48           49  Stanisław        Wójcik   
49           50    Enrique         Muñoz   
50           51     Joakim     Johansson   
51           52       Emma         Jones   
52           53       Phil        Hughes   
53           54      Steve        Murray   
54           55       Mark        Taylor   
55           56      Diego     Gutiérrez   
56           57       Luis         Rojas   
57           58      Manoj        Pareek   
58           59       Puja    Srivastava   

                                             company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                               None   
2                                               None   
3                                               None   
4                                   JetBrains s.r.o.   
5                                               None   
6                                               None   
7                                               None   
8                                               None   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                              None   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                              None   
18                                        Apple Inc.   
19                                              None   
20                                              None   
21                                              None   
22                                              None   
23                 

- The provided data represents a DataFrame containing information about customers.
- The DataFrame has multiple columns that provide details about each customer, including 'customer_id', 'first_name', 'last_name', 'company', 'address', 'city', 'state', 'country', 'postal_code', 'phone', 'fax', 'email', and 'support_rep_id'.
- The Customers DataFrame presents an extensive list of customers from various countries and cities around the world.
- Each customer is uniquely identified by their 'customer_id'.
- The DataFrame offers valuable insights into the demographic distribution of customers, including their names, contact information, and affiliations with companies or organizations.

In [8]:
query1 = pd.read_sql("""
SELECT *
FROM employee;
""", conn)

query1

employee_id last_name first_name                title  reports_to  \
0            1     Adams     Andrew      General Manager         NaN   
1            2   Edwards      Nancy        Sales Manager         1.0   
2            3   Peacock       Jane  Sales Support Agent         2.0   
3            4      Park   Margaret  Sales Support Agent         2.0   
4            5   Johnson      Steve  Sales Support Agent         2.0   
5            6  Mitchell    Michael           IT Manager         1.0   
6            7      King     Robert             IT Staff         6.0   
7            8  Callahan      Laura             IT Staff         6.0   

             birthdate            hire_date                      address  \
0  1962-02-18 00:00:00  2016-08-14 00:00:00          11120 Jasper Ave NW   
1  1958-12-08 00:00:00  2016-05-01 00:00:00                 825 8 Ave SW   
2  1973-08-29 00:00:00  2017-04-01 00:00:00                1111 6 Ave SW   
3  1947-09-19 00:00:00  2017-05-03 00:00:00             683 10 Street SW   
4  1965-03-03 00:00:00  2017-10-17 00:00:00                 7727B 41 Ave   
5  1973-07-01 00:00:00  2016-10-17 00:00:00         5827 Bowness Road NW   
6  1970-05-29 00:00:00  2017-01-02 00:00:00  590 Columbia Boulevard West   
7  1968-01-09 00:00:00  2017-03-04 00:00:00                  923 7 ST NW   

         city state country postal_code              phone                fax  \
0    Edmonton    AB  Canada     T5K 2N1  +1 (780) 428-9482  +1 (780) 428-3457   
1     Calgary    AB  Canada     T2P 2T3  +1 (403) 262-3443  +1 (403) 262-3322   
2     Calgary    AB  Canada     T2P 5M5  +1 (403) 262-3443  +1 (403) 262-6712   
3     Calgary    AB  Canada     T2P 5G3  +1 (403) 263-4423  +1 (403) 263-4289   
4     Calgary    AB  Canada     T3B 1Y7   1 (780) 836-9987   1 (780) 836-9543   
5     Calgary    AB  Canada     T3B 0C5  +1 (403) 246-9887  +1 (403) 246-9899   
6  Lethbridge    AB  Canada     T1K 5N8  +1 (403) 456-9986  +1 (403) 456-8485   
7  Lethbridge    AB  Canada     T1H 1Y8  +1 (403) 467-3351  +1 (403) 467-8772   

                      email  
0    andrew@chinookcorp.com  
1     nancy@chinookcorp.com  
2      jane@chinookcorp.com  
3  margaret@chinookcorp.com  
4     steve@chinookcorp.com  
5   michael@chinookcorp.com  
6    robert@chinookcorp.com  
7     laura@chinookcorp.com

- The provided data represents a DataFrame containing information about employees working at a company named "Chinook Corporation." The DataFrame consists of several columns such as 'employee_id', 'last_name', 'first_name', 'title', 'reports_to', 'birthdate', 'hire_date', 'address', 'city', 'state', 'country', 'postal_code', 'phone', 'fax', and 'email'.
- The Employees DataFrame presents a comprehensive overview of the staff employed at Chinook Corporation.
- Each employee is identified by their unique 'employee_id'. The DataFrame provides valuable details about the employees, including their full names, job titles, reporting managers, birthdates, hire dates, and contact information.

In [9]:
query1 = pd.read_sql("""
SELECT *
FROM genre;
""", conn)

query1

genre_id                name
0          1                Rock
1          2                Jazz
2          3               Metal
3          4  Alternative & Punk
4          5       Rock And Roll
5          6               Blues
6          7               Latin
7          8              Reggae
8          9                 Pop
9         10          Soundtrack
10        11          Bossa Nova
11        12      Easy Listening
12        13         Heavy Metal
13        14            R&B/Soul
14        15   Electronica/Dance
15        16               World
16        17         Hip Hop/Rap
17        18     Science Fiction
18        19            TV Shows
19        20    Sci Fi & Fantasy
20        21               Drama
21        22              Comedy
22        23         Alternative
23        24           Classical
24        25               Opera

The provided data represents a DataFrame containing information about music genres. The DataFrame consists of two columns: 'genre_id' and 'name'. Each row represents a different music genre, and the columns provide specific details about each genre.This DataFrame provides an extensive list of music genres, ranging from popular categories like Rock, Jazz, and Metal to more niche genres like Bossa Nova, Sci Fi & Fantasy, and Electronica/Dance. Music genres play a crucial role in categorizing and identifying different styles of music, allowing music enthusiasts to explore and enjoy a wide range of musical experiences.

In [11]:
query1 = pd.read_sql("""
SELECT *
FROM invoice;
""", conn)

query1

invoice_id  customer_id         invoice_date       billing_address  \
0             1           18  2017-01-03 00:00:00          627 Broadway   
1             2           30  2017-01-03 00:00:00      230 Elgin Street   
2             3           40  2017-01-05 00:00:00        8, Rue Hanovre   
3             4           18  2017-01-06 00:00:00          627 Broadway   
4             5           27  2017-01-07 00:00:00       1033 N Park Ave   
..          ...          ...                  ...                   ...   
609         610           55  2020-12-21 00:00:00     421 Bourke Street   
610         611           52  2020-12-27 00:00:00     202 Hoxton Street   
611         612           33  2020-12-27 00:00:00        5112 48 Street   
612         613           20  2020-12-29 00:00:00  541 Del Medio Avenue   
613         614           53  2020-12-30 00:00:00          113 Lupus St   

      billing_city billing_state billing_country billing_postal_code  total  
0         New York            NY             USA          10012-2612  15.84  
1           Ottawa            ON          Canada             K2P 1L7   9.90  
2            Paris          None          France               75002   1.98  
3         New York            NY             USA          10012-2612   7.92  
4           Tucson            AZ             USA               85719  16.83  
..             ...           ...             ...                 ...    ...  
609         Sidney           NSW       Australia                2010   6.93  
610         London          None  United Kingdom              N1 5LH   1.98  
611    Yellowknife            NT          Canada             X1A 1N6  11.88  
612  Mountain View            CA             USA           94040-111   8.91  
613         London          None  United Kingdom            SW1V 3EN  12.87  

[614 rows x 9 columns]

The provided data represents a DataFrame containing information about invoices. The DataFrame consists of several columns, including 'invoice_id', 'customer_id', 'invoice_date', 'billing_address', 'billing_city', 'billing_state', 'billing_country', 'billing_postal_code', and 'total'. Each row represents a unique invoice, and the columns provide specific details about each invoice and its associated customer.The DataFrame includes a total of 615 rows, each corresponding to a unique invoice. It presents a comprehensive overview of the sales transactions, detailing the customer's information, the purchase date, the billing address, and the total amount spent on each transaction

In [12]:
query1 = pd.read_sql("""
SELECT *
FROM invoice_line;
""", conn)

query1

invoice_line_id  invoice_id  track_id  unit_price  quantity
0                   1           1      1158        0.99         1
1                   2           1      1159        0.99         1
2                   3           1      1160        0.99         1
3                   4           1      1161        0.99         1
4                   5           1      1162        0.99         1
...               ...         ...       ...         ...       ...
4752             4753         614      2659        0.99         1
4753             4754         614      2660        0.99         1
4754             4755         614      2661        0.99         1
4755             4756         614      2662        0.99         1
4756             4757         614      2663        0.99         1

[4757 rows x 5 columns]

The provided data represents a DataFrame containing information about invoice line items. The DataFrame consists of several columns, including 'invoice_line_id', 'invoice_id', 'track_id', 'unit_price', and 'quantity'. Each row represents a unique line item within an invoice, and the columns provide specific details about each line item and the associated invoice.The DataFrame includes a total of 4,757 rows, each corresponding to a unique invoice line item. It provides a comprehensive breakdown of the individual music tracks purchased by customers in each invoice. The unit price and quantity information allow businesses to calculate the total revenue generated from each invoice and understand customer preferences regarding specific music tracks.

In [13]:
query1 = pd.read_sql("""
SELECT *
FROM media_type;
""", conn)

query1

media_type_id                         name
0              1              MPEG audio file
1              2     Protected AAC audio file
2              3  Protected MPEG-4 video file
3              4     Purchased AAC audio file
4              5               AAC audio file

The provided data represents a DataFrame containing information about media types. The DataFrame consists of two columns: 'media_type_id' and 'name'. Each row represents a different media type, and the columns provide specific details about each media type.Media Type ID_column contains unique identifiers for each media type.
Name_column provides the name or description of each media type.Media types are essential for categorizing different types of media files, such as audio and video files. They help in organizing and identifying media content, ensuring compatibility with different devices and software applications.

In [14]:
query1 = pd.read_sql("""
SELECT *
FROM playlist;
""", conn)

query1

playlist_id                        name
0             1                       Music
1             2                      Movies
2             3                    TV Shows
3             4                  Audiobooks
4             5                  90’s Music
5             6                  Audiobooks
6             7                      Movies
7             8                       Music
8             9                Music Videos
9            10                    TV Shows
10           11             Brazilian Music
11           12                   Classical
12           13   Classical 101 - Deep Cuts
13           14  Classical 101 - Next Steps
14           15  Classical 101 - The Basics
15           16                      Grunge
16           17         Heavy Metal Classic
17           18                 On-The-Go 1

The provided data represents a DataFrame containing information about playlists. The DataFrame consists of two columns: 'playlist_id' and 'name'. Each row represents a different playlist, and the columns provide specific details about each playlist.Playlists are a way to group and organize media content, such as music tracks, movies, TV shows, audiobooks, and more. Users can create playlists based on their preferences, moods, or specific genres, allowing them to easily access and enjoy their favorite content

In [15]:
query1 = pd.read_sql("""
SELECT *
FROM playlist_track;
""", conn)

query1

playlist_id  track_id
0               1      3402
1               1      3389
2               1      3390
3               1      3391
4               1      3392
...           ...       ...
8710           17      2094
8711           17      2095
8712           17      2096
8713           17      3290
8714           18       597

[8715 rows x 2 columns]

The provided data represents a DataFrame containing information about the association between playlists and tracks. The DataFrame consists of two columns: 'playlist_id' and 'track_id'. Each row represents a track's inclusion in a specific playlist, and the columns provide details about the relationship between playlists and tracks.The data shows that multiple tracks are associated with some playlists. For example, playlist ID 1 contains several tracks with track IDs 3402, 3389, 3390, 3391, 3392, and more. Playlist ID 17 also includes multiple tracks with track IDs 2094, 2095, 2096, 3290, and more. On the other hand, playlist ID 18 contains a single track with track ID 597.

In [16]:
query1 = pd.read_sql("""
SELECT *
FROM track;
""", conn)

query1

track_id                                               name  album_id  \
0            1            For Those About To Rock (We Salute You)         1   
1            2                                  Balls to the Wall         2   
2            3                                    Fast As a Shark         3   
3            4                                  Restless and Wild         3   
4            5                               Princess of the Dawn         3   
...        ...                                                ...       ...   
3498      3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...       343   
3499      3500  String Quartet No. 12 in C Minor, D. 703 "Quar...       344   
3500      3501               L'orfeo, Act 3, Sinfonia (Orchestra)       345   
3501      3502  Quintet for Horn, Violin, 2 Violas, and Cello ...       346   
3502      3503                                      Koyaanisqatsi       347   

      media_type_id  genre_id  \
0                 1         1   
1                 2         1   
2                 2         1   
3                 2         1   
4                 2         1   
...             ...       ...   
3498              2        24   
3499              2        24   
3500              2        24   
3501              2        24   
3502              2        10   

                                               composer  milliseconds  \
0             Angus Young, Malcolm Young, Brian Johnson        343719   
1                                                  None        342562   
2     F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...        230619   
3     F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...        252051   
4                            Deaffy & R.A. Smith-Diesel        375418   
...                                                 ...           ...   
3498                                               None        286741   
3499                                     Franz Schubert        139200   
3500                                 Claudio Monteverdi         66639   
3501                            Wolfgang Amadeus Mozart        221331   
3502                                       Philip Glass        206005   

         bytes  unit_price  
0     11170334        0.99  
1      5510424        0.99  
2      3990994        0.99  
3      4331779        0.99  
4      6290521        0.99  
...        ...         ...  
3498   4718950        0.99  
3499   2283131        0.99  
3500   1189062        0.99  
3501   3665114        0.99  
3502   3305164        0.99  

[3503 rows x 9 columns]

The provided data represents a DataFrame containing information about music tracks in a music database. The DataFrame consists of several columns, each providing specific details about each track.The dataset contains information about 3503 music tracks in total. Each row represents a different music track, and the columns provide specific details about each track, including its name, album, media type, genre, composer (if available), duration, size, and unit price.The data is crucial for organizing and managing music content in a music database or music store. It allows users to search, filter, and purchase tracks based on various criteria such as genre, composer, album, and price.

# Segment 2: Track/Album Analysis

How many tracks does Chinook store have in total?

In [17]:
Q2 = pd.read_sql("""
SELECT count(track_id) 
FROM Track;""",conn)
Q2

count(track_id)
0             3503

Which tracks are the most popular among customers?

In [22]:
Q3 = pd.read_sql("""
SELECT t.track_id, 
t.name AS track_name, 
COUNT(il.track_id) AS Popular
FROM track t
JOIN invoice_line il ON t.track_id = il.track_id
GROUP BY t.track_id, t.name
ORDER BY Popular DESC limit 10;""",conn)
Q3

track_id                track_name  Popular
0      3336                  War Pigs       31
1      1489      Are You Experienced?       14
2      1495             Highway Chile       14
3         6     Put The Finger On You       13
4      1487  Third Stone From The Sun       13
5      1490                   Hey Joe       13
6      1483         Love Or Confusion       12
7      2558               Radio/Video       12
8      1129           Dead And Broken       11
9      1479                 Foxy Lady       11

What are the top-selling albums in the Chinook store?

In [29]:
Q4 = pd.read_sql("""
SELECT a.Album_Id, 
a.title AS Album_Title, 
COUNT(il.track_id) AS Top_Selling_Albums
FROM 
  album a
JOIN track t ON a.album_id = t.album_id
JOIN invoice_line il ON t.track_id = il.track_id
GROUP BY a.album_id, a.title
ORDER BY Top_Selling_Albums DESC;""",conn)
Q4


album_id                                        Album_Title  \
0         120                               Are You Experienced?   
1          88                                           Faceless   
2         207                                          Mezmerize   
3         119                                           Get Born   
4         214                                          The Doors   
..        ...                                                ...   
246       302                     Mascagni: Cavalleria Rusticana   
247       314                                English Renaissance   
248       325                   Bartok: Violin & Viola Concertos   
249       329                             South American Getaway   
250       336  Prokofiev: Symphony No.5 & Stravinksy: Le Sacr...   

     Top_Selling_Albums  
0                   187  
1                    96  
2                    93  
3                    90  
4                    83  
..                  ...  
246                   1  
247                   1  
248                   1  
249                   1  
250                   1  

[251 rows x 3 columns]

Which artist or album has the highest number of tracks in the store?

In [34]:
Q5 = pd.read_sql("""
SELECT
    a.Artist_Id, a.Name AS Artist_Name, al.Album_Id,
    al.title AS Album_title,
    COUNT(t.track_id) AS T_Tracks
FROM 
    Artist a
JOIN 
    Album al ON a.Artist_id = al.Artist_id
JOIN 
    Track t ON al.Album_id = t.Album_id
GROUP BY 
    a.Artist_id, a.Name, al.Album_id, al.Title
ORDER BY 
    T_Tracks DESC
LIMIT 10;""",conn)

Q5


artist_id                     Artist_Name  album_id  \
0        100                   Lenny Kravitz       141   
1         17                   Chico Buarque        23   
2         81                    Eric Clapton        73   
3        149                            Lost       229   
4        149                            Lost       230   
5        156                      The Office       251   
6         85                   Frank Sinatra        83   
7        149                            Lost       231   
8        158  Battlestar Galactica (Classic)       253   
9         18     Chico Science & Nação Zumbi        24   

                                  Album_title  T_Tracks  
0                               Greatest Hits        57  
1                              Minha Historia        34  
2                                   Unplugged        30  
3                              Lost, Season 3        26  
4                              Lost, Season 1        25  
5                        The Office, Season 3        25  
6  My Way: The Best Of Frank Sinatra [Disc 1]        24  
7                              Lost, Season 2        24  
8    Battlestar Galactica (Classic), Season 1        24  
9                              Afrociberdelia        23

Determine the number of tracks for each album and calculate the total duration of each album. Identify the albums that have the highest average track duration.

In [10]:
Q6= pd.read_sql("""
select
  a.Album_id, a.Title as Album_Title,count(t.track_id),
  sum(t.Milliseconds)/60000 as Duration_Album,
    round(AVG(t.Milliseconds),2)/60000 AS AverageTrack_Milliseconds
from album a
    left join track t on a.album_id = t.album_id
group by 1,2
order by 5 desc;

""",conn)
Q6

album_id                                    Album_Title  \
0         253       Battlestar Galactica (Classic), Season 1   
1         227                 Battlestar Galactica, Season 3   
2         229                                 Lost, Season 3   
3         231                                 Lost, Season 2   
4         226         Battlestar Galactica: The Story So Far   
..        ...                                            ...   
342       328  Charpentier: Divertissements, Airs & Concerts   
343       314                            English Renaissance   
344       318                       SCRIABIN: Vers la flamme   
345       345                            Monteverdi: L'Orfeo   
346       340    Liszt - 12 Études D'Execution Transcendante   

     count(t.track_id)  Duration_Album  AverageTrack_Milliseconds  
0                   24            1170                  48.759572  
1                   19             879                  46.304422  
2                   26            1177                  45.298450  
3                   24            1054                  43.951133  
4                    1              43                  43.704167  
..                 ...             ...                        ...  
342                  1               1                   1.837767  
343                  2               3                   1.691350  
344                  1               1                   1.688217  
345                  1               1                   1.110650  
346                  1               0                   0.863000  

[347 rows x 5 columns]

How does the genre distribution vary among albums?

In [11]:
Q7= pd.read_sql("""SELECT
    Album.album_id,
    Album.title AS Album_Title,
    Genre.name AS Genre_Name,
    COUNT(Track.track_id) AS Num_Tracks
FROM
    Album
JOIN Track ON Album.album_id = Track.album_id
JOIN Genre ON Track.genre_id = Genre.genre_id
GROUP BY
    Album.album_id, Album.title, Genre.name
ORDER BY
    Album.album_id, Genre.name;

""",conn)
Q7


album_id                                        Album_Title  Genre_Name  \
0           1              For Those About To Rock We Salute You        Rock   
1           2                                  Balls to the Wall        Rock   
2           3                                  Restless and Wild        Rock   
3           4                                  Let There Be Rock        Rock   
4           5                                           Big Ones        Rock   
..        ...                                                ...         ...   
355       343                             Respighi:Pines of Rome   Classical   
356       344  Schubert: The Late String Quartets & String Qu...   Classical   
357       345                                Monteverdi: L'Orfeo   Classical   
358       346                              Mozart: Chamber Music   Classical   
359       347  Koyaanisqatsi (Soundtrack from the Motion Pict...  Soundtrack   

     Num_Tracks  
0            10  
1             1  
2             3  
3             8  
4            15  
..          ...  
355           1  
356           1  
357           1  
358           1  
359           1  

[360 rows x 4 columns]

# Segment 3: Customer Analysis

Analyse customer demographics, purchase history, and preferences

In [4]:
Q8 = pd.read_sql("""
select *
from customer
""", conn)
Q8

customer_id first_name     last_name  \
0             1       Luís     Gonçalves   
1             2     Leonie        Köhler   
2             3   François      Tremblay   
3             4      Bjørn        Hansen   
4             5  František   Wichterlová   
5             6     Helena          Holý   
6             7     Astrid        Gruber   
7             8       Daan       Peeters   
8             9       Kara       Nielsen   
9            10    Eduardo       Martins   
10           11  Alexandre         Rocha   
11           12    Roberto       Almeida   
12           13   Fernanda         Ramos   
13           14       Mark       Philips   
14           15   Jennifer      Peterson   
15           16      Frank        Harris   
16           17       Jack         Smith   
17           18   Michelle        Brooks   
18           19        Tim         Goyer   
19           20        Dan        Miller   
20           21      Kathy         Chase   
21           22    Heather       Leacock   
22           23       John        Gordon   
23           24      Frank       Ralston   
24           25     Victor       Stevens   
25           26    Richard    Cunningham   
26           27    Patrick          Gray   
27           28      Julia       Barnett   
28           29     Robert         Brown   
29           30     Edward       Francis   
30           31     Martha          Silk   
31           32      Aaron      Mitchell   
32           33      Ellie      Sullivan   
33           34       João     Fernandes   
34           35   Madalena       Sampaio   
35           36     Hannah     Schneider   
36           37       Fynn    Zimmermann   
37           38     Niklas      Schröder   
38           39    Camille       Bernard   
39           40  Dominique      Lefebvre   
40           41       Marc        Dubois   
41           42      Wyatt        Girard   
42           43   Isabelle       Mercier   
43           44      Terhi    Hämäläinen   
44           45   Ladislav        Kovács   
45           46       Hugh      O'Reilly   
46           47      Lucas       Mancini   
47           48   Johannes  Van der Berg   
48           49  Stanisław        Wójcik   
49           50    Enrique         Muñoz   
50           51     Joakim     Johansson   
51           52       Emma         Jones   
52           53       Phil        Hughes   
53           54      Steve        Murray   
54           55       Mark        Taylor   
55           56      Diego     Gutiérrez   
56           57       Luis         Rojas   
57           58      Manoj        Pareek   
58           59       Puja    Srivastava   

                                             company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                               None   
2                                               None   
3                                               None   
4                                   JetBrains s.r.o.   
5                                               None   
6                                               None   
7                                               None   
8                                               None   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                              None   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                              None   
18                                        Apple Inc.   
19                                              None   
20                                              None   
21                                              None   
22                                              None   
23                 

How many unique customers are there in the Chinook store?

In [12]:
Q9 = pd.read_sql("""
SELECT COUNT(DISTINCT customer_id) AS unique_customers
FROM customer;
""", conn)
Q9

unique_customers
0                59

What is the distribution of customers across different countries?


In [21]:
Q10 = pd.read_sql("""
SELECT Country, COUNT(customer_id) AS T_Customers
FROM Customer
GROUP BY Country
ORDER BY T_Customers DESC limit 10;
""",conn)
Q10

country  T_Customers
0             USA           13
1          Canada            8
2          France            5
3          Brazil            5
4         Germany            4
5  United Kingdom            3
6        Portugal            2
7           India            2
8  Czech Republic            2
9          Sweden            1

Identify the average number of tracks purchased per customer for each genre


In [26]:
Q11= pd.read_sql("""
SELECT g.name AS Genre,
AVG(il.quantity) AS Avg_TPurchased
FROM invoice_line il
JOIN track t ON il.track_id = t.track_id
JOIN Genre g ON t.Genre_id = g.Genre_id
GROUP BY g.name
ORDER BY Avg_TPurchased DESC limit 10;

""",conn)
Q11

Genre  Avg_TPurchased
0     TV Shows             1.0
1   Soundtrack             1.0
2         Rock             1.0
3       Reggae             1.0
4     R&B/Soul             1.0
5          Pop             1.0
6        Metal             1.0
7        Latin             1.0
8         Jazz             1.0
9  Hip Hop/Rap             1.0

What is the average purchase value per customer?

In [35]:
Q12 = pd.read_sql("""
SELECT 
    c.customer_id,
    c.first_name || ' ' || c.last_name AS customer_name,
    COUNT(i.invoice_id) AS total_purchases,
    SUM(i.total) AS total_purchase_value,
    AVG(i.total) AS average_purchase_value
FROM customer c
LEFT JOIN invoice i ON c.customer_id = i.customer_id
GROUP BY c.customer_id, customer_name
ORDER BY average_purchase_value DESC;
""",conn)
Q12

customer_id          customer_name  total_purchases  total_purchase_value  \
0             3      François Tremblay                9                 99.99   
1             6            Helena Holý               12                128.70   
2            29           Robert Brown                4                 40.59   
3            18        Michelle Brooks                8                 79.20   
4            37        Fynn Zimmermann               10                 94.05   
5            27           Patrick Gray                9                 84.15   
6            16           Frank Harris                8                 74.25   
7            42           Wyatt Girard               11                 99.99   
8            24          Frank Ralston                8                 71.28   
9            50          Enrique Muñoz               11                 98.01   
10           53            Phil Hughes               11                 98.01   
11           59        Puja Srivastava                8                 71.28   
12           46          Hugh O'Reilly               13                114.84   
13           39        Camille Bernard                9                 79.20   
14           54           Steve Murray                9                 79.20   
15           32         Aaron Mitchell                8                 70.29   
16            8           Daan Peeters                7                 60.39   
17           58           Manoj Pareek               13                111.87   
18           52             Emma Jones                8                 68.31   
19            1         Luís Gonçalves               13                108.90   
20           21            Kathy Chase               11                 91.08   
21           17             Jack Smith               12                 98.01   
22           38        Niklas Schröder                9                 73.26   
23           55            Mark Taylor               10                 81.18   
24            4           Bjørn Hansen                9                 72.27   
25            5  František Wichterlová               18                144.54   
26           40     Dominique Lefebvre                9                 72.27   
27           34         João Fernandes               13                102.96   
28           56        Diego Gutiérrez                5                 39.60   
29           20             Dan Miller               12                 95.04   
30           45        Ladislav Kovács               10                 78.21   
31           36       Hannah Schneider               11                 85.14   
32            7          Astrid Gruber                9                 69.30   
33           22        Heather Leacock               12                 92.07   
34           25         Victor Stevens               10                 76.23   
35           49       Stanisław Wójcik               10                 76.23   
36           51       Joakim Johansson               10                 75.24   
37            2          Leonie Köhler               11                 82.17   
38           12        Roberto Almeida               11                 82.17   
39           57             Luis Rojas               13                 97.02   
40           15      Jennifer Peterson                9                 66.33   
41           28          Julia Barnett               10                 72.27   
42           44       Terhi Hämäläinen               11                 79.20   
43           26     Richard Cunningham               12                 86.13   
44           41            Marc Dubois                9                 64.35   
45           13         Fernanda Ramos               15                106.92   
46           30         Edward Francis               13                 91.08   
47           11        Alexandre Rocha               10                 69.30   
48           23            John Gor

Which customers have made the highest number of purchases?

In [41]:
Q13= pd.read_sql("""
SELECT
    c.customer_id,
    c.first_name || ' ' || c.last_name AS customer_name,
    COUNT(*) AS total_purchases
FROM customer c
JOIN invoice i ON c.customer_id = i.customer_id
GROUP BY c.customer_id, customer_name
ORDER BY total_purchases DESC
LIMIT 10;
""",conn)
Q13


customer_id          customer_name  total_purchases
0            5  František Wichterlová               18
1           35       Madalena Sampaio               16
2           13         Fernanda Ramos               15
3            1         Luís Gonçalves               13
4           30         Edward Francis               13
5           34         João Fernandes               13
6           46          Hugh O'Reilly               13
7           57             Luis Rojas               13
8           58           Manoj Pareek               13
9            6            Helena Holý               12

# Segment 4: Support analysis

What is the total sales revenue generated by each sales support agent?


In [43]:
Q14 = pd.read_sql("""
SELECT
e.employee_id,e.first_name || ' ' || e.last_name AS sales_support_agent,
SUM(i.total) AS total_sales_revenue
FROM employee e
JOIN customer c ON e.employee_id = c.support_rep_id
JOIN invoice i ON c.customer_id = i.customer_id
GROUP BY e.employee_id, sales_support_agent
ORDER BY total_sales_revenue DESC;
""",conn)
Q14


employee_id sales_support_agent  total_sales_revenue
0            3        Jane Peacock              1731.51
1            4       Margaret Park              1584.00
2            5       Steve Johnson              1393.92

What is the average order value for each sales support agent?

In [50]:
Q15 = pd.read_sql("""
SELECT
    e.employee_id,
    e.first_name || ' ' || e.last_name AS sales_agent_name,
    COUNT(i.invoice_id) AS total_orders,
    SUM(i.total) AS total_sales,
    AVG(i.total) AS average_order_value
FROM employee e
LEFT JOIN customer c ON e.employee_id = c.support_rep_id
LEFT JOIN invoice i ON c.customer_id = i.customer_id
WHERE e.title = 'Sales Support Agent'
GROUP BY e.employee_id, sales_agent_name
ORDER BY average_order_value DESC;
""",conn)
Q15

employee_id sales_agent_name  total_orders  total_sales  \
0            3     Jane Peacock           212      1731.51   
1            5    Steve Johnson           188      1393.92   
2            4    Margaret Park           214      1584.00   

   average_order_value  
0             8.167500  
1             7.414468  
2             7.401869

Is there any significant difference in sales performance between agents reporting to different managers?

In [52]:
Q16 = pd.read_sql("""
SELECT
    m.employee_id AS manager_id,
    m.first_name || ' ' || m.last_name AS manager_name,
    e.employee_id AS agent_id,
    e.first_name || ' ' || e.last_name AS agent_name,
    COUNT(i.invoice_id) AS total_orders,
    SUM(i.total) AS total_sales,
    AVG(i.total) AS average_order_value
FROM employee e
LEFT JOIN customer c ON e.employee_id = c.support_rep_id
LEFT JOIN invoice i ON c.customer_id = i.customer_id
JOIN employee m ON e.reports_to = m.employee_id
WHERE e.title = 'Sales Support Agent'
GROUP BY m.employee_id, manager_name, e.employee_id, agent_name
ORDER BY manager_id, average_order_value DESC;

""",conn)
Q16


manager_id   manager_name  agent_id     agent_name  total_orders  \
0           2  Nancy Edwards         3   Jane Peacock           212   
1           2  Nancy Edwards         5  Steve Johnson           188   
2           2  Nancy Edwards         4  Margaret Park           214   

   total_sales  average_order_value  
0      1731.51             8.167500  
1      1393.92             7.414468  
2      1584.00             7.401869

How does the hiring date of sales support agents impact their sales performance?



In [53]:
Q17 = pd.read_sql("""
SELECT
    DATE(e.hire_date) AS hiring_date,
    COUNT(i.invoice_id) AS total_orders,
    SUM(i.total) AS total_sales,
    AVG(i.total) AS average_order_value
FROM employee e
LEFT JOIN customer c ON e.employee_id = c.support_rep_id
LEFT JOIN invoice i ON c.customer_id = i.customer_id
WHERE e.title = 'Sales Support Agent'
GROUP BY DATE(e.hire_date)
ORDER BY DATE(e.hire_date);
""",conn)
Q17

hiring_date  total_orders  total_sales  average_order_value
0  2017-04-01           212      1731.51             8.167500
1  2017-05-03           214      1584.00             7.401869
2  2017-10-17           188      1393.92             7.414468

# Segment 5: Sales Analysis

What is the total revenue generated by the Chinook store?

In [54]:
Q18 = pd.read_sql("""
SELECT SUM(total) AS total_revenue
FROM invoice;
""",conn)
Q18

total_revenue
0        4709.43

How does the revenue vary across different countries?

In [55]:
Q19 = pd.read_sql("""SELECT 
    billing_country AS country,
    SUM(total) AS total_revenue
FROM invoice
GROUP BY billing_country
ORDER BY total_revenue DESC;
""",conn)
Q19

country  total_revenue
0              USA        1040.49
1           Canada         535.59
2           Brazil         427.68
3           France         389.07
4          Germany         334.62
5   Czech Republic         273.24
6   United Kingdom         245.52
7         Portugal         185.13
8            India         183.15
9          Ireland         114.84
10           Spain          98.01
11           Chile          97.02
12       Australia          81.18
13         Finland          79.20
14         Hungary          78.21
15          Poland          76.23
16          Sweden          75.24
17          Norway          72.27
18         Austria          69.30
19     Netherlands          65.34
20         Belgium          60.39
21           Italy          50.49
22       Argentina          39.60
23         Denmark          37.62

Calculate the percentage of total sales contributed by each genre in the USA and identify the best-selling genres and artists.

In [60]:
Q20 = pd.read_sql("""SELECT
    g.name AS genre,
    a.name AS artist,
    COUNT(t.track_id) AS track_count,
    ROUND(SUM(il.unit_price), 2) AS total_sales,
    ROUND((SUM(il.unit_price) / (SELECT SUM(total) FROM invoice WHERE billing_country = 'USA')) * 100, 2) AS percentage_of_total_sales
FROM
    invoice i
JOIN invoice_line il ON i.invoice_id = il.invoice_id
JOIN track t ON il.track_id = t.track_id
JOIN genre g ON t.genre_id = g.genre_id
JOIN album al ON t.album_id = al.album_id
JOIN artist a ON al.artist_id = a.artist_id
WHERE
    i.billing_country = 'USA'
GROUP BY
    g.name, a.name
ORDER BY
    total_sales DESC;

""",conn)
Q20

genre              artist  track_count  total_sales  \
0                  Rock           Van Halen           43        42.57   
1                  Rock  The Rolling Stones           37        36.63   
2                  Rock             Nirvana           35        34.65   
3    Alternative & Punk           Green Day           32        31.68   
4    Alternative & Punk              R.E.M.           32        31.68   
..                  ...                 ...          ...          ...   
111               Metal     Bruce Dickinson            1         0.99   
112                Rock       Faith No More            1         0.99   
113          Soundtrack        Gilberto Gil            1         0.99   
114          Soundtrack          Passengers            1         0.99   
115            TV Shows                Lost            1         0.99   

     percentage_of_total_sales  
0                         4.09  
1                         3.52  
2                         3.33  
3                         3.04  
4                         3.04  
..                         ...  
111                       0.10  
112                       0.10  
113                       0.10  
114                       0.10  
115                       0.10  

[116 rows x 5 columns]

Evaluate sales performance over time to identify seasonal trends and peak periods

In [62]:
Q21 = pd.read_sql("""SELECT
    strftime('%Y-%m', i.invoice_date) AS Date,
    SUM(i.total) AS total_sales
FROM
    invoice i
GROUP BY
    Date
ORDER BY
    Date;
""",conn)
Q21

Date  total_sales
0   2017-01       126.72
1   2017-02       141.57
2   2017-03       103.95
3   2017-04       142.56
4   2017-05       104.94
5   2017-06        75.24
6   2017-07       108.90
7   2017-08        88.11
8   2017-09       107.91
9   2017-10        79.20
10  2017-11        94.05
11  2017-12        28.71
12  2018-01       183.15
13  2018-02       116.82
14  2018-03       148.50
15  2018-04        63.36
16  2018-05        76.23
17  2018-06        92.07
18  2018-07        85.14
19  2018-08       106.92
20  2018-09        56.43
21  2018-10        78.21
22  2018-11        42.57
23  2018-12        98.01
24  2019-01        85.14
25  2019-02        59.40
26  2019-03       125.73
27  2019-04       116.82
28  2019-05       104.94
29  2019-06        92.07
30  2019-07       124.74
31  2019-08       159.39
32  2019-09       126.72
33  2019-10        43.56
34  2019-11        71.28
35  2019-12       111.87
36  2020-01        43.56
37  2020-02        97.02
38  2020-03        78.21
39  2020-04       119.79
40  2020-05        82.17
41  2020-06       120.78
42  2020-07        76.23
43  2020-08        72.27
44  2020-09        95.04
45  2020-10       144.54
46  2020-11        83.16
47  2020-12       125.73

Determine the average transaction value and customer lifetime value to gauge business profitability

In [71]:
Q22 = pd.read_sql("""SELECT
    AVG(total) AS customer_lifetime_value
FROM
    (
        SELECT
            customer_id,
            SUM(total) AS total
        FROM
            invoice
        GROUP BY
            customer_id
    ) AS customer_sales;
    """,conn)
Q22

customer_lifetime_value
0                79.820847

# Segment 6: SQL Concepts

Use subqueries to find the top-selling tracks in the USA and identify their associated genres

In [73]:
Q23 = pd.read_sql("""
SELECT
    t.track_id,
    t.name AS track_name,
    g.name AS genre_name,
    COUNT(il.track_id) AS total_sales
FROM
    invoice i
JOIN
    invoice_line il ON i.invoice_id = il.invoice_id
JOIN
    track t ON il.track_id = t.track_id
JOIN
    genre g ON t.genre_id = g.genre_id
WHERE
    i.billing_country = 'USA'
GROUP BY
    t.track_id
ORDER BY
    total_sales DESC
LIMIT 10;
     """,conn)
Q23

track_id                                     track_name   genre_name  \
0      3336                                       War Pigs  Alternative   
1      3465  You Know I'm No Good (feat. Ghostface Killah)     R&B/Soul   
2      2647                               End Of The Night         Rock   
3      2646                                I Looked At You         Rock   
4      2560                            Violent Pornography        Metal   
5      1995                           Scentless Apprentice         Rock   
6      1495                                  Highway Chile         Rock   
7       153                                     Evil Woman        Metal   
8        13                       Night Of The Long Knives         Rock   
9      3293                                Big City Nights         Rock   

   total_sales  
0            6  
1            5  
2            4  
3            4  
4            4  
5            4  
6            4  
7            4  
8            4  
9            3

Implement joins to retrieve customer information along with their preferred genres and country of origin

In [74]:
Q24 = pd.read_sql("""
SELECT
    c.customer_id,
    c.first_name || ' ' || c.last_name AS customer_name,
    c.country,
    g.name AS preferred_genre
FROM
    customer c
LEFT JOIN
    invoice i ON c.customer_id = i.customer_id
LEFT JOIN
    invoice_line il ON i.invoice_id = il.invoice_id
LEFT JOIN
    track t ON il.track_id = t.track_id
LEFT JOIN
    genre g ON t.genre_id = g.genre_id
GROUP BY
    c.customer_id
ORDER BY
    c.customer_id;
        """,conn)
Q24

customer_id          customer_name         country     preferred_genre
0             1         Luís Gonçalves          Brazil                Rock
1             2          Leonie Köhler         Germany                Rock
2             3      François Tremblay          Canada                Rock
3             4           Bjørn Hansen          Norway                Rock
4             5  František Wichterlová  Czech Republic  Alternative & Punk
5             6            Helena Holý  Czech Republic                Rock
6             7          Astrid Gruber         Austria                Rock
7             8           Daan Peeters         Belgium            R&B/Soul
8             9           Kara Nielsen         Denmark   Electronica/Dance
9            10        Eduardo Martins          Brazil  Alternative & Punk
10           11        Alexandre Rocha          Brazil                Rock
11           12        Roberto Almeida          Brazil                Rock
12           13         Fernanda Ramos          Brazil                Rock
13           14           Mark Philips          Canada                Rock
14           15      Jennifer Peterson          Canada  Alternative & Punk
15           16           Frank Harris             USA                Rock
16           17             Jack Smith             USA                Rock
17           18        Michelle Brooks             USA                Rock
18           19              Tim Goyer             USA                Rock
19           20             Dan Miller             USA                Rock
20           21            Kathy Chase             USA                Rock
21           22        Heather Leacock             USA               Blues
22           23            John Gordon             USA               Metal
23           24          Frank Ralston             USA                Rock
24           25         Victor Stevens             USA  Alternative & Punk
25           26     Richard Cunningham             USA            R&B/Soul
26           27           Patrick Gray             USA                Rock
27           28          Julia Barnett             USA                Rock
28           29           Robert Brown          Canada                Rock
29           30         Edward Francis          Canada                 Pop
30           31            Martha Silk          Canada      Easy Listening
31           32         Aaron Mitchell          Canada            R&B/Soul
32           33         Ellie Sullivan          Canada                Jazz
33           34         João Fernandes        Portugal                Rock
34           35       Madalena Sampaio        Portugal                Rock
35           36       Hannah Schneider         Germany               Metal
36           37        Fynn Zimmermann         Germany                Rock
37           38        Niklas Schröder         Germany  Alternative & Punk
38           39        Camille Bernard          France                Rock
39           40     Dominique Lefebvre          France                Rock
40           41            Marc Dubois          France      Easy Listening
41           42           Wyatt Girard          France   Electronica/Dance
42           43       Isabelle Mercier          France               Blues
43           44       Terhi Hämäläinen         Finland              Reggae
44           45        Ladislav Kovács         Hungary                Rock
45           46          Hugh O'Reilly         Ireland               Blues
46           47          Lucas Mancini           Italy  Alternative & Punk
47           48  Johannes Van der Berg     Netherlands            R&B/Soul
48           49       Stanisław Wójcik          Poland                 Pop
49           50          Enrique Muñoz           Spain               Metal
50           51       Joakim Johansson          Sweden                Rock
51           52             Emma Jones  United Kingdom                Rock
52           53            Ph

Apply window functions to rank genres based on their sales performance in the USA

In [75]:
Q25 = pd.read_sql( """
SELECT
    g.name AS genre,
    SUM(il.unit_price * il.quantity) AS total_sales,
    RANK() OVER (ORDER BY SUM(il.unit_price * il.quantity) DESC) AS genre_rank
FROM
    genre g
JOIN
    track t ON g.genre_id = t.genre_id
JOIN
    invoice_line il ON t.track_id = il.track_id
JOIN
    invoice i ON il.invoice_id = i.invoice_id
WHERE
    i.billing_country = 'USA'
GROUP BY
    g.name;
        """,conn)
Q25

genre  total_sales  genre_rank
0                 Rock       555.39           1
1   Alternative & Punk       128.70           2
2                Metal       122.76           3
3             R&B/Soul        52.47           4
4                Blues        35.64           5
5          Alternative        34.65           6
6                  Pop        21.78           7
7                Latin        21.78           7
8          Hip Hop/Rap        19.80           9
9                 Jazz        13.86          10
10      Easy Listening        12.87          11
11              Reggae         5.94          12
12   Electronica/Dance         4.95          13
13           Classical         3.96          14
14         Heavy Metal         2.97          15
15          Soundtrack         1.98          16
16            TV Shows         0.99          17

Create stored procedures to calculate the average revenue per sale for each genre in the USA.

In [76]:
Q26 = pd.read_sql( """
SELECT
        g.name AS genre,
        AVG(il.unit_price * il.quantity) AS average_revenue_per_sale
    FROM
        genre g
    JOIN
        track t ON g.genre_id = t.genre_id
    JOIN
        invoice_line il ON t.track_id = il.track_id
    JOIN
        invoice i ON il.invoice_id = i.invoice_id
    WHERE
        i.billing_country = 'USA'
    GROUP BY
        g.name;
        """,conn)
Q26

genre  average_revenue_per_sale
0          Alternative                      0.99
1   Alternative & Punk                      0.99
2                Blues                      0.99
3            Classical                      0.99
4       Easy Listening                      0.99
5    Electronica/Dance                      0.99
6          Heavy Metal                      0.99
7          Hip Hop/Rap                      0.99
8                 Jazz                      0.99
9                Latin                      0.99
10               Metal                      0.99
11                 Pop                      0.99
12            R&B/Soul                      0.99
13              Reggae                      0.99
14                Rock                      0.99
15          Soundtrack                      0.99
16            TV Shows                      0.99

# Segment 7: Market Expansion and Advertising Strategy

Recommend the three albums from the new record label that should be prioritised for advertising and promotion in the USA based on genre sales analysis.

In [78]:
Q27 = pd.read_sql( """
SELECT
    a.album_id,
    a.title AS album_title,
    g.name AS genre_name,
    SUM(il.unit_price * il.quantity) AS total_revenue
FROM
    album a
JOIN
    track t ON a.album_id = t.album_id
JOIN
    genre g ON t.genre_id = g.genre_id
JOIN
    invoice_line il ON t.track_id = il.track_id
JOIN
    invoice i ON il.invoice_id = i.invoice_id
WHERE
    i.billing_country = 'USA'
GROUP BY
    a.album_id, album_title, genre_name
ORDER BY
    total_revenue DESC;
        """,conn)
Q27

album_id                                        album_title   genre_name  \
0         120                               Are You Experienced?         Rock   
1         163         From The Muddy Banks Of The Wishkah [live]         Rock   
2         214                                          The Doors         Rock   
3         146  Seek And Shall Find: More Of The Best (1963-1981)     R&B/Soul   
4          72                               The Cream Of Clapton        Blues   
..        ...                                                ...          ...   
169       278                             Bach: The Cello Suites    Classical   
170       293                        Pavarotti's Opera Made Easy    Classical   
171       304     Beethoven Piano Sonatas: Moonlight & Pastorale    Classical   
172       314                                English Renaissance    Classical   
173       323              Carried to Dust (Bonus Track Version)  Alternative   

     total_revenue  
0            27.72  
1            27.72  
2            26.73  
3            25.74  
4            24.75  
..             ...  
169           0.99  
170           0.99  
171           0.99  
172           0.99  
173           0.99  

[174 rows x 4 columns]

Identify high-value customers and develop personalised marketing strategies to nurture customer loyalty

In [79]:
Q28 = pd.read_sql( """
SELECT
    c.customer_id,
    c.first_name || ' ' || c.last_name AS customer_name,
    COUNT(i.invoice_id) AS total_purchases,
    SUM(i.total) AS total_purchase_value
FROM
    customer c
JOIN
    invoice i ON c.customer_id = i.customer_id
GROUP BY
    c.customer_id, customer_name
ORDER BY
    total_purchase_value DESC;
   """,conn)
Q28

customer_id          customer_name  total_purchases  total_purchase_value
0             5  František Wichterlová               18                144.54
1             6            Helena Holý               12                128.70
2            46          Hugh O'Reilly               13                114.84
3            58           Manoj Pareek               13                111.87
4             1         Luís Gonçalves               13                108.90
5            13         Fernanda Ramos               15                106.92
6            34         João Fernandes               13                102.96
7             3      François Tremblay                9                 99.99
8            42           Wyatt Girard               11                 99.99
9            17             Jack Smith               12                 98.01
10           50          Enrique Muñoz               11                 98.01
11           53            Phil Hughes               11                 98.01
12           57             Luis Rojas               13                 97.02
13           20             Dan Miller               12                 95.04
14           37        Fynn Zimmermann               10                 94.05
15           22        Heather Leacock               12                 92.07
16           21            Kathy Chase               11                 91.08
17           30         Edward Francis               13                 91.08
18           26     Richard Cunningham               12                 86.13
19           36       Hannah Schneider               11                 85.14
20           27           Patrick Gray                9                 84.15
21            2          Leonie Köhler               11                 82.17
22           12        Roberto Almeida               11                 82.17
23           35       Madalena Sampaio               16                 82.17
24           55            Mark Taylor               10                 81.18
25           18        Michelle Brooks                8                 79.20
26           39        Camille Bernard                9                 79.20
27           44       Terhi Hämäläinen               11                 79.20
28           54           Steve Murray                9                 79.20
29           45        Ladislav Kovács               10                 78.21
30           25         Victor Stevens               10                 76.23
31           49       Stanisław Wójcik               10                 76.23
32           33         Ellie Sullivan               12                 75.24
33           51       Joakim Johansson               10                 75.24
34           16           Frank Harris                8                 74.25
35           38        Niklas Schröder                9                 73.26
36           43       Isabelle Mercier               12                 73.26
37            4           Bjørn Hansen                9                 72.27
38           28          Julia Barnett               10                 72.27
39           40     Dominique Lefebvre                9                 72.27
40           24          Frank Ralston                8                 71.28
41           59        Puja Srivastava                8                 71.28
42           32         Aaron Mitchell                8                 70.29
43            7          Astrid Gruber                9                 69.30
44           11        Alexandre Rocha               10                 69.30
45           52             Emma Jones                8                 68.31
46           15      Jennifer Peterson                9                 66.33
47           23            John Gordon               10                 66.33
48           48  Johannes Van der Berg               10                 65.34
49           41            Marc Dubois                9                 64.35
50           31           

Analyse the potential for expanding sales of different genres in other countries based on their sales performance in the USA.

In [81]:
Q29 = pd.read_sql( """
SELECT
        g.name AS genre,
        SUM(il.unit_price * il.quantity) AS total_usa_sales
    FROM
        invoice_line il
    JOIN
        track t ON il.track_id = t.track_id
    JOIN
        genre g ON t.genre_id = g.genre_id
    JOIN
        invoice i ON il.invoice_id = i.invoice_id
    WHERE
        i.billing_country = 'USA'
    GROUP BY
        genre;
           """,conn)
Q29

genre  total_usa_sales
0          Alternative            34.65
1   Alternative & Punk           128.70
2                Blues            35.64
3            Classical             3.96
4       Easy Listening            12.87
5    Electronica/Dance             4.95
6          Heavy Metal             2.97
7          Hip Hop/Rap            19.80
8                 Jazz            13.86
9                Latin            21.78
10               Metal           122.76
11                 Pop            21.78
12            R&B/Soul            52.47
13              Reggae             5.94
14                Rock           555.39
15          Soundtrack             1.98
16            TV Shows             0.99

Determine the top-selling genres in countries other than the USA and identify any commonalities or differences.

In [82]:
Q30 = pd.read_sql( """
SELECT
        i.billing_country,
        g.name AS genre,
        SUM(il.unit_price * il.quantity) AS total_sales
    FROM
        invoice_line il
    JOIN
        track t ON il.track_id = t.track_id
    JOIN
        genre g ON t.genre_id = g.genre_id
    JOIN
        invoice i ON il.invoice_id = i.invoice_id
    WHERE
        i.billing_country <> 'USA'
    GROUP BY
        i.billing_country, genre;
               """,conn)
Q30

billing_country               genre  total_sales
0         Argentina         Alternative         0.99
1         Argentina  Alternative & Punk        16.83
2         Argentina               Blues         1.98
3         Argentina      Easy Listening         0.99
4         Argentina         Heavy Metal         0.99
..              ...                 ...          ...
252  United Kingdom               Metal        30.69
253  United Kingdom                 Pop         0.99
254  United Kingdom            R&B/Soul         1.98
255  United Kingdom              Reggae         1.98
256  United Kingdom                Rock       164.34

[257 rows x 3 columns]